### PR TITLE
Add SwitchSequencer

### PR DESCRIPTION
**Pull request by ImJeremyHe**

### CODE REVIEW

File: execution/gethexec/switch_sequencer.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+	return s.espresso.Start(ctx)
+}
+
+func (s *SwitchSequencer) SwitchToCentralized(ctx context.Context) error {
+	if s.mode == SequencingMode_Centralized {
+		return nil
+	}
+	s.mode = SequencingMode_Centralized
+	s.espresso.StopAndWait()
+	return s.espresso.Start(ctx)
```
Comment (jbearer) [inline execution/gethexec/switch_sequencer.go:67]:
```suggestion
	return s.centralized.Start(ctx)
```

File: execution/gethexec/switch_sequencer.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+func (s *SwitchSequencer) Start(ctx context.Context) error {
+	err := s.getRunningSequencer().Start(ctx)
+	if err != nil {
+		return err
+	}
+	s.CallIteratively(func(ctx context.Context) time.Duration {
+		now := time.Now()
+		if s.IsRunningEspressoMode() {
+			if s.espresso.lastCreated.Add(s.maxHotshotDirftTime).Before(now) {
```
Comment (jbearer) [inline execution/gethexec/switch_sequencer.go:97]: We may need to change the trigger for switching to centralized mode. Will discuss in Zulip

File: execution/gethexec/switch_sequencer.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+)
+
+type SwitchSequencer struct {
+	stopwaiter.StopWaiter
+
+	centralized *Sequencer
+	espresso    *EspressoSequencer
+
+	mode                     int
+	maxHotshotDirftTime      time.Duration
```
Comment (sveitser) [inline execution/gethexec/switch_sequencer.go:24]:
```suggestion
	maxHotshotDriftTime      time.Duration
```

File: execution/gethexec/switch_sequencer.go [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+
+func (s *SwitchSequencer) IsRunningEspressoMode() bool {
+	return s.mode == SequencingMode_Espresso
+}
+
+func (s *SwitchSequencer) SwitchToEspresso(ctx context.Context) error {
+	if s.IsRunningEspressoMode() {
+		return nil
+	}
+	log.Info("Switching to espresso sequencer")
```
Comment (nomaxg) [inline execution/gethexec/switch_sequencer.go:68]: Can we log the drift time and exactly how far we have drifted when we switch?

File: execution/gethexec/switch_sequencer.go [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+	s.centralized.StopAndWait()
+	return s.espresso.Start(ctx)
+}
+
+func (s *SwitchSequencer) SwitchToCentralized(ctx context.Context) error {
+	if !s.IsRunningEspressoMode() {
+		return nil
+	}
+	s.mode = SequencingMode_Centralized
+	log.Info("Switching to centrialized sequencer")
```
Comment (nomaxg) [inline execution/gethexec/switch_sequencer.go:81]: Same here, let's log precisely why we are switching

File: execution/gethexec/switch_sequencer.go [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+}
+
+func (s *SwitchSequencer) Start(ctx context.Context) error {
+	s.StopWaiter.Start(ctx, s)
+	err := s.getRunningSequencer().Start(ctx)
+	if err != nil {
+		return err
+	}
+
+	if s.lightClient != nil {
```
Comment (nomaxg) [inline execution/gethexec/switch_sequencer.go:118]: Log a message like "light client not specified, we will only be sequencing mode"

**Review by nomaxg — APPROVED** [review]

**Pull request merged**